csrc/utils.h

```diff
@@ -621,86 +623,205 @@ void checkAllEqual(std::initializer_list<T> elements) {
 
 #if __cplusplus >= 202302L
 
+using std::views::enumerate;
 using std::views::zip;
 
 #else
 
 namespace views {
-
-template <std::ranges::input_range... Rs>
+template <std::ranges::view... Rs>
 class zip_view : public std::ranges::view_interface<zip_view<Rs...>> {
  private:
   std::tuple<Rs...> bases;
 
-  // Iterator for begin()
-  struct begin_iterator {
+  static constexpr bool is_bidirectional =
+      (std::ranges::bidirectional_range<Rs> && ...);
+
+  struct iterator {
     std::tuple<std::ranges::iterator_t<Rs>...> iterators;
 
     using value_type = std::tuple<std::ranges::range_value_t<Rs>...>;
     using reference = std::tuple<std::ranges::range_reference_t<Rs>...>;
     using difference_type = std::ptrdiff_t;
+    using iterator_category = std::conditional_t<
```

Comment:
Is this used anywhere?

Reply:
This is not used in our codebase, but it does belong to the [LegacyIterator](https://en.cppreference.com/w/cpp/named_req/Iterator) way to make iterator traits work (we are not using iterator traits yet, but we may use it in the future). This is not very modern, but not deprecated either, and for this specific case, it takes less line of code compared to the modern way. Considering that this is just a backport of C++23, I would just use whatever works and wait for 3 years to upgrade to C++23 and completely abandon this code. So I am not trying to making the code here perfect.